dids/diddht/internal/dns/did.go

```diff
@@ -0,0 +1,277 @@
+	"golang.org/x/net/dns/dnsmessage"
+)
+
+// MarshalDIDDocument packs a DID document into a TXT DNS resource records and adds to the DNS message Answers
+func MarshalDIDDocument(d *didcore.Document) ([]byte, error) {
+
+	// create root record
+	var msg dnsmessage.Message
+
+	// get sorted VM IDs
+	sortedIDs := pluckSort(d.VerificationMethod)
+	vmIDToK := map[string]string{}
+
+	vmBEP44Keys := []string{}
+
+	for k, id := range sortedIDs {
+		_k := fmt.Sprintf("k%d", k)
+		vmIDToK[id] = _k
+		vmBEP44Keys = append(vmBEP44Keys, _k)
+	}
+
+	sToK := map[string]string{}
+	sKeys := []string{}
+	for k, v := range d.Service {
+		_k := fmt.Sprintf("s%d", k)
+		sToK[v.ID] = _k
+		sKeys = append(sKeys, _k)
+	}
+
+	rootProps := map[string][]string{
+		"v":    {"1"},
+		"id":   {d.ID},
+		"vm":   vmBEP44Keys,
+		"auth": methodsToKeys(d.Authentication, vmIDToK),
+		"asm":  methodsToKeys(d.AssertionMethod, vmIDToK),
+		"agm":  methodsToKeys(d.KeyAgreement, vmIDToK),
+		"inv":  methodsToKeys(d.CapabilityInvocation, vmIDToK),
+		"del":  methodsToKeys(d.CapabilityDelegation, vmIDToK),
+		"srv":  sKeys,
+		"cnt":  d.Controller,
+		"aka":  d.AlsoKnownAs,
+	}
+
+	rootPropsSerialized := []string{}
+	for k, v := range rootProps {
+		if len(v) == 0 {
+			continue
+		}
+		prop := fmt.Sprintf("%s=%s", k, strings.Join(v, ","))
+		rootPropsSerialized = append(rootPropsSerialized, prop)
+	}
+
+	resource, err := newResource("_did.", strings.Join(rootPropsSerialized, ";"))
+	if err != nil {
+		return nil, err
+	}
+	msg.Answers = append(msg.Answers, resource)
+
+	// add verification methods to dns message
+	for _, v := range d.VerificationMethod {
+		vm := v
+		// look for the key after the # in the verification method ID
+		key, ok := vmIDToK[v.ID]
+		if !ok {
+			// TODO handle error
+			continue
+		}
+		buf, err := MarshalVerificationMethod(&vm)
+		if err != nil {
+			return nil, err
+		}
+
+		resource, err := newResource(fmt.Sprintf("_%s._did.", key), buf)
+
+		if err != nil {
+			return nil, err
+		}
+
+		msg.Answers = append(msg.Answers, resource)
+	}
+
+	// add services to dns message
+	for _, s := range d.Service {
+		key, ok := sToK[s.ID]
+		if !ok {
+			// TODO handle error
+			continue
+		}
+		if err := MarshalService(key, s, &msg); err != nil {
+			return nil, err
+		}
+	}
+
+	msgByes, err := msg.Pack()
+	if err != nil {
+		return nil, err
+	}
+
+	return msgByes, nil
+}
+
+// UnmarshalDIDDocument unpacks the TXT DNS resource records and returns a DID document
+func UnmarshalDIDDocument(didID string, payload []byte) (*didcore.Document, error) {
+	decoder, err := parseDNSDID(payload)
+	if err != nil {
+		return nil, err
+	}
+
+	doc, err := decoder.DIDDocument(didID)
+	if err != nil {
+		return nil, err
+	}
+
+	return doc, nil
+}
+
+// MarshalVerificationMethod packs a verification method into a TXT DNS resource record and adds to the DNS message Answers
+func MarshalVerificationMethod(vm *didcore.VerificationMethod) (string, error) {
```

Comment:
FWIW I'm much more sympathetic to using the "marshal" language here b/c it's networking-adjacent (being to/from DNS resource)